src/gradient/eigen_value_gradient.py

```diff
@@ -2,21 +2,28 @@
 
 
 class EigenValueGradient:
-    def __init__(self, ngm_small_tensor: torch.Tensor, dominant_eig_vec: torch.Tensor):
-        self.ngm_small_tensor = ngm_small_tensor
-        self.dominant_eig_vec = dominant_eig_vec
+    def __init__(self, ngm_small_grads: torch.Tensor, left_eigen_vec: torch.Tensor,
+                 right_eigen_vec: torch.Tensor):
 
-        self.eig_val_cm_grad = None
+        self.left_eig_vec = left_eigen_vec
+        self.right_eig_vec = right_eigen_vec
+        self.ngm_small_grads = ngm_small_grads
 
-    def run(self, ngm_small_grads: torch.Tensor):
-        # Reshape self.dominant_eig_vec to be a column vector
-        eig_vec = self.dominant_eig_vec.view(-1, 1)  # Shape: [16, 1]
+        self.r0_cm_grad = None
 
-        # Compute A @ x: ngm_small_grads @ eig_vec for each slice
-        a_dot_x = torch.matmul(ngm_small_grads, eig_vec).squeeze(dim=2)  # Shape: [16, 136]
+    def run(self):
+        # Calculate v^T.w
+        normalization = torch.dot(self.left_eig_vec, self.right_eig_vec)
+        # calculate v.w^T / v^T.w
+        s_ij = self.left_eig_vec.view(-1, 1) * self.right_eig_vec.view(1, -1) / \
+               normalization
 
-        # Compute x.T @ A @ x
-        # Reshape x.T to match the multiplication with B
-        # get the eig_vec_transpose
-        eig_vec_transpose = self.dominant_eig_vec.T
-        self.eig_val_cm_grad = torch.matmul(eig_vec_transpose, a_dot_x)
+        # Add extra dimension to align for broadcasting
+        s_ij = s_ij.unsqueeze(-1)  # Shape: (16, 16, 1)
+
+        weighted_grads = torch.matmul(self.ngm_small_grads, s_ij)  # Shape: (16, 136, 1)
```

Comment:
This should be `self.ngm_small_grads * s_ij` or something similar.
You should have a `16 x 16` matrix (`s_ij`), which is ELEMENTWISE MULTIPLIED by a `16 x 136 x 16` (`grads`) matrix along dim 0 and 2 - my guess is, that you need a `16 x 1 x 16` format (for `s_ij`). Then you will get a `16 x 136 x 16` matrix, for which you do the summation along dimensions 0 and 2.